src/QGCApplication.cc

```diff
@@ -553,6 +553,11 @@ bool QGCApplication::_initForNormalAppBoot()
     QSettings settings;
 
     _qmlAppEngine = toolbox()->corePlugin()->createQmlApplicationEngine(this);
+
+    // 3D Viewer initialization
+    _qgcViewer3D = new QGCViewer3D(this);
+    _qgcViewer3D->initQml(_qmlAppEngine, this);
```

Comment:
So this is probably related to my comment above about the module thing. Can you explain to me what this QGCViewer3D class does. Probably need a high level overview of the architecture, class hierarchy and how it functions to make sure this gets integrated correctly.

Reply:
This class instantiates and initializes all the classes required for the 3D viewer. I created this class just to put all the requirements of the 3D viewer in one place, making it easier to debug or edit without affecting other parts of QGC. However, this class can be removed and its content can be merged with the "QGCApplication" class.

In general, for the moment, the  "QGCViewer3D" class contains three essential classes: 
- OsmParser: to load the OSM map file.
- QmlBackend: for linking the QML and C++.
- MetaDataStreamer: to save and load the user's custom settings for the 3D viewer, such as the path of the OSM map file and the height bias set by the user.

Reply:
So basically this is to make the c++ classes available to Qml code correct. If so you should change this to work like the other C++ classes exposed to Qml already in QGC. Look at QGCApplication::_initCommon. You want to register you classes using qmlRegisterType with your own module name. Then you use Import in Qml and create them as needed in your Qml code. Does that make sense?

Reply:
In my latest commit today, I have already added some lines of code in the QGCApplication::_initCommon function to register my C++ types using qmlRegisterType and then I imported them in qml flies as you mentioned.

But the QGCViewer3D class does a different job. It makes the instantiated C++ variable visible to qml, and for this, it requires the variable "_qmlAppEngine" which is initialized in QGCApplication::_initForNormalAppBoot(). So, the suitable place to instantiate and initialize QGCViewer3D is inside QGCApplication::_initForNormalAppBoot().

Reply:
Sorry for the multiple replies.
Please let me add more details about QGCViewer3D class. 

The line of code "_qgcViewer3D = new QGCViewer3D(this);" can be placed anywhere in QGCApplication.cc. However, the function "_qgcViewer3D->initQml(_qmlAppEngine, this);" should be inside "QGCApplication::_initForNormalAppBoot()", after initializing "_qmlAppEngine" and before calling "toolbox()->corePlugin()->createRootWindow(_qmlAppEngine);", since the variables required for the 3D viewer should be initialized before creating the RootWindow. 
Please let me know if that is ok.

Reply:
You shouldn't be injecting this into all Qml at a global level. That isn't the way QGC does things. If you look at FactPanelController on both the C++ and Qml side of thing you should see how you make a C++ class/object available to and/or instantiated from the Qml side of the code. 

Basically the way this should work:
* You have worker C++ classes you need to make the Qml work
* You register those class(es) as creatable from Qml using qmlRegisterType in QGCApplication
* Then in your Qml you create them in there

This way:
* They are not global things that magically show up in all Qml
* They only are live when they are needed

Reply:
Thanks for the explanation. Yes, that makes sense. 
So, in my new commit as you suggested, I have registered all my types/classes with qmlRegisterType and there is no global declaration anymore. I created a new Class Viewer3DFacts to handle accessing C++ variables and classes from the QML side. Please let me know if this looks good to you.